File: heapsort.c
#include <stdio.h>
#include <stdlib.h>
#include <windows.h>

void heapify(int arr[], int n, int i) {
    int largest = i;
    int l = 2*i + 1;
    int r = 2*i + 2;

    if (l < n && arr[l] > arr[largest])
        largest = l;
    if (r < n && arr[r] > arr[largest])
        largest = r;

    if (largest != i) {
        int temp = arr[i];
        arr[i] = arr[largest];
        arr[largest] = temp;
        heapify(arr, n, largest);
    }
}

void heapSort(int arr[], int n) {
    for (int i = n/2 - 1; i >= 0; i--)
        heapify(arr, n, i);

    for (int i = n - 1; i >= 0; i--) {
        int temp = arr[0];
        arr[0] = arr[i];
        arr[i] = temp;
        heapify(arr, i, 0);
    }
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        printf("Uso: %s <tamanho>\n", argv[0]);
        return 1;
    }

    int n = atoi(argv[1]);
    int *arr = malloc(n * sizeof(int));

    for (int i = 0; i < n; i++)
        arr[i] = rand();

    LARGE_INTEGER start, end, freq;
    QueryPerformanceFrequency(&freq);
    QueryPerformanceCounter(&start);

    heapSort(arr, n);

    QueryPerformanceCounter(&end);

    double time_taken = (double)(end.QuadPart - start.QuadPart) / freq.QuadPart;

    printf("%f\n", time_taken);

    free(arr);
    return 0;
}

File: heapsort.py
import time
import random
import sys

def heapify(arr, n, i):
    largest = i
    l = 2*i + 1
    r = 2*i + 2
    
    if l < n and arr[l] > arr[largest]:
        largest = l
    if r < n and arr[r] > arr[largest]:
        largest = r
    
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest)

def heap_sort(arr):
    n = len(arr)
    for i in range(n//2 - 1, -1, -1):
        heapify(arr, n, i)
    for i in range(n - 1, 0, -1):
        arr[0], arr[i] = arr[i], arr[0]
        heapify(arr, i, 0)

if len(sys.argv) != 2:
    print("Uso: python heapsort.py <tamanho>")
    sys.exit()

n = int(sys.argv[1])
arr = [random.randint(0, 10**9) for _ in range(n)]

start = time.perf_counter()
heap_sort(arr)
end = time.perf_counter()

print(f"{end - start:.6f}")
